src/extract_gpuinfo_tpu.c

```diff
@@ -0,0 +1,290 @@
+  char name[8];
+  int64_t device_id;
+  int64_t memory_usage;
+  int64_t total_memory;
+  double duty_cycle_pct;
+  int64_t pid;
+};
+
+static bool gpuinfo_tpu_init(void);
+static void gpuinfo_tpu_shutdown(void);
+static const char *gpuinfo_tpu_last_error_string(void);
+static bool gpuinfo_tpu_get_device_handles(struct list_head *devices, unsigned *count);
+static void gpuinfo_tpu_populate_static_info(struct gpu_info *_gpu_info);
+static void gpuinfo_tpu_refresh_dynamic_info(struct gpu_info *_gpu_info);
+static void gpuinfo_tpu_get_running_processes(struct gpu_info *_gpu_info);
+static bool is_cache_valid(void);
+static bool refresh_tpu_cache(void);
+static void reset_tpu_cache(bool);
+static void free_ptr(void **ptr);
+
+struct gpu_vendor gpu_vendor_tpu = {
+    .init = gpuinfo_tpu_init,
+    .shutdown = gpuinfo_tpu_shutdown,
+    .last_error_string = gpuinfo_tpu_last_error_string,
+    .get_device_handles = gpuinfo_tpu_get_device_handles,
+    .populate_static_info = gpuinfo_tpu_populate_static_info,
+    .refresh_dynamic_info = gpuinfo_tpu_refresh_dynamic_info,
+    .refresh_running_processes = gpuinfo_tpu_get_running_processes,
+    .name = "TPU",
+};
+
+__attribute__((constructor)) static void init_extract_gpuinfo_tpu(void) { 
+  register_gpu_vendor(&gpu_vendor_tpu);
+}
+
+int64_t tpu_chip_count = -1;
+static struct gpu_info_tpu *gpu_infos;
+
+#define STRINGIFY(x) STRINGIFY_HELPER_(x)
+#define STRINGIFY_HELPER_(x) #x
+
+#define VENDOR_TPU 0x1ae0
+#define VENDOR_TPU_STR STRINGIFY(VENDOR_TPU)
+
+#define MAX(x, y) ((x >= y) ? (x) : (y))
+#define MIN(x, y) ((x <= y) ? (x) : (y))
+
+#define int64 long long
+
+int (*_tpu_chip_count)(void);
+int (*_tpu_metrics)(int port, int64 *device_ids, int64 *memory_usage, 
+                    int64 *total_memory, double *duty_cycle_pct, int n);
+int (*_tpu_pids)(int64 *pids, int n);
+
+char *libname = "libtpuinfo.so";
+// -1 means allowing libtpuinfo to select the default port
+// env LIBTPUINFO_GRPC_PORT={int} allows setting the port via an environment variable
+// $ env LIBTPUINFO_GRPC_PORT=8431 nvtop
+int tpu_runtime_monitoring_port = -1;  
+
+/* TPU info cache ------------------------------------------------------------------------------- */
+struct tpu_chip_usage_data *latest_chips_usage_data = NULL;
+nvtop_time last_cache_refresh;
+int64 *_pids, *_device_ids, *_memory_usage, *_total_memory;
+double* _duty_cycle_pct;
+
+bool is_cache_valid(void) {
+  nvtop_time current_time;
+  nvtop_get_current_time(&current_time);
+  uint64_t t_diff_ns = nvtop_difftime_u64(last_cache_refresh, current_time);
+  return t_diff_ns < 900 * 1000 * 1000; // 900ms
+}
+
+bool refresh_tpu_cache(void) {
+  if (is_cache_valid()) return true;
+  nvtop_get_current_time(&last_cache_refresh);
+  if (tpu_chip_count <= 0) return false;
+  if (_tpu_pids(_pids, tpu_chip_count) != 0) {
+    reset_tpu_cache(false);
+    return false;
+  }
+  for (int i = 0; i < tpu_chip_count; i++) latest_chips_usage_data[i].pid = _pids[i];
```

Comment:
tpu_chip_count is int64_t so better use these everywhere as induction variables or move tpu_chip_count to using int (having more than 2billion TPUs on the same machine is unlikely, I guess 🙃)

Reply:
Sounds good, I'll move to `int64_t`